book/src/dev/rfcs/0010-v5-transaction.md

```diff
@@ -0,0 +1,224 @@
+- `orchard transaction version` - Transactions that support orchard data. Currently only V5.
+- `sapling transaction version` - Transactions that support sapling data. Currently V4 and V5 but the data is implemented differently in them.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+V5 transactions are described by the protocol in the second table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus).
+
+We need the raw representation to Serialize/Deserialize the transaction however Zebra uses its own representation internally.
+
+V5 transactions are the only ones that will support orchard transactions with `Orchard` data types.
+
+V4 and V5 transactions both support sapling data but they are implemented differently. By this reason we need to split sapling data types into V4 and V5. We do this by using `enum` data types.
+
+The order of some of the fields changed from V4 to V5. For example the `lock_time` and `expiry_height` were moved above the transparent inputs and outputs.
+
+All of the changes proposed in this document are only to the `zebra-chain` crate.
+
+To highlight changes all document comments from the code snippets in the [reference section](#reference-level-explanation) were removed.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We start by looking how a V4 (already implemented) transaction is represented in Zebra.
+
+This transaction version is specified by the protocol in the first table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus).
+
+
+```
+V4 {
+    inputs: Vec<transparent::Input>,
+    outputs: Vec<transparent::Output>,
+    lock_time: LockTime,
+    expiry_height: block::Height,
+    value_balance: Amount,
+    joinsplit_data: Option<JoinSplitData<Groth16Proof>>,
+    shielded_data: Option<ShieldedData>,
+}
+```
+
+Currently the `ShieldedData` type is defined in `zebra-chain/src/transaction/shielded_data.rs` as follows:
+
+```
+pub struct ShieldedData {
+    pub first: Either<Spend, Output>,
+    pub rest_spends: Vec<Spend>,
+    pub rest_outputs: Vec<Output>,
+    pub binding_sig: Signature<Binding>,
+}
+```
+
+We know by protocol (2nd table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus)) that V5 transactions will support sapling data however we also know by protocol that spends ([Spend Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#spendencodingandconsensus), See †) and outputs ([Output Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#outputencodingandconsensus), See †) fields change from V4 to V5.
+
+Here we have the proposed changes for V4 transactions:
+
+```
+V4 {
+    ...
+    sapling_value_balance: Amount,
+    ...
+    sapling_shielded_data: Option<SaplingShieldedData::V4>,
```

Comment:
Rust doesn't allow us to specify a single enum variant as a generic parameter.

So a `Transaction::V4` could contain `SaplingShieldedData::V5`, containing a mixed vector or `Spend::V4` and `Spend::V5`.

There are two different ways to prevent this kind of bug:
1. Check for the correct variant during serialization at runtime, and write proptests that do deserialization/serialization round-trips

2. Encode the differences between the variants as types, and pass those types as generic parameters

For example, the `ShieldedData` type could look like:
```rust
type NoAnchor = ();

pub enum ShieldedData {
    V4 {
        pub first: Either<Spend<tree::Root>, Output>,
        pub rest_spends: Vec<Spend<tree::Root>>,
        ....
    },
    V5 {
        pub first: Either<Spend<NoAnchor>, Output>,
        pub rest_spends: Vec<Spend<NoAnchor>>,
        pub anchor: tree::Root,
        ...
    }
```

@oxarbitrage @dconnolly what do you think?

I can do up a design for this later today, after I've done a pass for some other minor fixes.

Reply:
This is now commit d860cf5d4a28698548819276ffc2d8ab0d98a7f3